File: server/app/services/gemini_service.py
import google.generativeai as genai
from ..core.config import settings
import json

class GeminiService:
    def __init__(self):
        genai.configure(api_key=settings.GEMINI_API_KEY)
        self.model = genai.GenerativeModel('gemini-2.5-flash') # Using gemini-2.5-flash as requested

        # Define the tools (Function Declarations)
        self.tools = [
            {
                "name": "write_file",
                "description": "Writes content to a file.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "file_path": { "type": "string" },
                        "content": { "type": "string" }
                    },
                    "required": ["file_path", "content"]
                }
            },
            {
                "name": "read_file",
                "description": "Reads the content of a file.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "file_path": { "type": "string" }
                    },
                    "required": ["file_path"]
                }
            },
            {
                "name": "run_shell_command",
                "description": "Executes a shell command.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "command": { "type": "string" }
                    },
                    "required": ["command"]
                }
            },
            {
                "name": "finish",
                "description": "Indicates task completion.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "message": { "type": "string" }
                    },
                    "required": ["message"]
                }
            }
        ]

    async def generate_response(self, history: list) -> str:
        try:
            response = await self.model.generate_content_async(history, tools=self.tools)
        except Exception as e:
            # Fallback without tools if there's an error
            print(f"Error with tools: {e}")
            response = await self.model.generate_content_async(history)
            # Return a default action if no tools available
            return '{"action": "finish", "message": "Error with function calling. Please try again."}'
        # Parse the functionCall from the response
        if response.candidates and response.candidates[0].content.parts[0].function_call:
            function_call = response.candidates[0].content.parts[0].function_call
            return json.dumps({"action": function_call.name, **function_call.args})
        else:
            # If no function call, return the text response
            return response.text

gemini_service = GeminiService()

File: server-ts/src/services/GeminiService.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { ToolRegistry } from '../tools/ToolRegistry';
import { GeminiResponse } from '../models/GeminiResponse';

export class GeminiService {
    private genAI: GoogleGenerativeAI;
    private model: any;
    private toolRegistry: ToolRegistry;

    constructor(apiKey: string, toolRegistry: ToolRegistry) {
        if (!apiKey) {
            throw new Error('GEMINI_API_KEY is required');
        }

        this.genAI = new GoogleGenerativeAI(apiKey);
        this.toolRegistry = toolRegistry;
        
        // Configure the model
        this.model = this.genAI.getGenerativeModel({
            model: 'gemini-2.5-flash',
            generationConfig: {
                temperature: 0.1,
            },
        });
    }

    public async generateResponse(history: any[]): Promise<GeminiResponse> {
        try {
            // Add system prompt with available tools
            const systemPrompt = this.buildSystemPrompt();
            const fullHistory = [
                {
                    role: 'user',
                    parts: [{ text: systemPrompt }]
                },
                ...history
            ];

            const result = await this.model.generateContent({
                contents: fullHistory,
            });

            const response = result.response;
            const text = response.text();

            // Parse JSON response
            try {
                const parsedResponse = JSON.parse(text);
                return {
                    conversation_id: '', // Will be set by controller
                    action: parsedResponse.action || 'finish',
                    message: parsedResponse.message,
                    command: parsedResponse.command,
                    file_path: parsedResponse.file_path,
                    content: parsedResponse.content
                };
            } catch (parseError) {
                return {
                    conversation_id: '',
                    action: 'finish',
                    message: `Invalid JSON response from Gemini: ${text}`
                };
            }
        } catch (error: any) {
            return {
                conversation_id: '',
                action: 'finish',
                message: `Error: ${error.message}`
            };
        }
    }

    private buildSystemPrompt(): string {
        const availableTools = this.toolRegistry.getAvailableTools();
        
        return `You are an AI software engineering agent helping with GitHub issues.

Available tools:
${availableTools.map(tool => `- ${tool}`).join('\n')}

You must respond with valid JSON in one of these formats:

For shell commands:
{"action": "run_shell_command", "command": "your_command_here"}

For reading files:
{"action": "read_file", "file_path": "path/to/file"}

For writing files:
{"action": "write_file", "file_path": "path/to/file", "content": "file_content"}

For task completion:
{"action": "finish", "message": "Summary of what was accomplished"}

Always analyze the issue carefully and use the appropriate tools to complete the task.
When you're done, use the 'finish' action with a summary of what you accomplished.`;
    }
}